common/asyncworkflow/queue/consumer/default_consumer.go

```diff
@@ -0,0 +1,228 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package consumer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/uber/cadence/.gen/go/sqlblobs"
+	"github.com/uber/cadence/client/frontend"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/codec"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/messaging"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/types"
+)
+
+const (
+	defaultShutdownTimeout = 5 * time.Second
+	defaultConcurrency     = 100
+)
+
+type defaultConsumer struct {
+	queueID         string
+	innerConsumer   messaging.Consumer
+	logger          log.Logger
+	scope           metrics.Scope
+	frontendClient  frontend.Client
+	ctx             context.Context
+	cancelFn        context.CancelFunc
+	wg              sync.WaitGroup
+	shutdownTimeout time.Duration
+	msgDecoder      codec.BinaryEncoder
+	concurrency     int
+}
+
+type Option func(*defaultConsumer)
+
+func WithConcurrency(concurrency int) Option {
+	return func(c *defaultConsumer) {
+		c.concurrency = concurrency
+	}
+}
+
+func New(
+	queueID string,
+	innerConsumer messaging.Consumer,
+	logger log.Logger,
+	metricsClient metrics.Client,
+	frontendClient frontend.Client,
+	options ...Option,
+) *defaultConsumer {
+	ctx, cancelFn := context.WithCancel(context.Background())
+	c := &defaultConsumer{
+		queueID:         queueID,
+		innerConsumer:   innerConsumer,
+		logger:          logger.WithTags(tag.Dynamic("queue-id", queueID)),
+		scope:           metricsClient.Scope(metrics.AsyncWorkflowConsumerScope),
+		frontendClient:  frontendClient,
+		ctx:             ctx,
+		cancelFn:        cancelFn,
+		shutdownTimeout: defaultShutdownTimeout,
+		msgDecoder:      codec.NewThriftRWEncoder(),
+		concurrency:     defaultConcurrency,
+	}
+
+	for _, opt := range options {
+		opt(c)
+	}
+
+	return c
+}
+
+func (c *defaultConsumer) Start() error {
+	if err := c.innerConsumer.Start(); err != nil {
+		return err
+	}
+
+	for i := 0; i < c.concurrency; i++ {
+		c.wg.Add(1)
+		go c.runProcessLoop()
+		c.logger.Info("Started process loop", tag.Counter(i))
+	}
+	c.logger.Info("Started consumer", tag.Dynamic("concurrency", c.concurrency))
+	return nil
+}
+
+func (c *defaultConsumer) Stop() {
+	c.logger.Info("Stopping consumer")
+	c.cancelFn()
+	c.wg.Wait()
+	if !common.AwaitWaitGroup(&c.wg, c.shutdownTimeout) {
+		c.logger.Warn("Consumer timed out on shutdown", tag.Dynamic("timeout", c.shutdownTimeout))
+		return
+	}
+
+	c.innerConsumer.Stop()
+	c.logger.Info("Stopped consumer")
+}
+
+func (c *defaultConsumer) runProcessLoop() {
+	defer c.wg.Done()
+
```

Comment:
panic recover?

Reply:
I'm not planning to handle panics here. there's no expected code path that should raise panic so an actual panic should cause crash. If we had a legacy underlying library/code that were raising panics instead of errors then it would make sense to handle it here.

Reply:
I guess crashing the worker's not the worst thing in the world. I get quite afraid of not handling panics with recovery in history, matching and frontend for their tendency to cause cascading failures (host crashes, the shard moves to a new host, repeat)